<a href="https://colab.research.google.com/github/glgunderson/INFOB2DA-PA2/blob/main/pa4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Dashboard Visualizations and Coordinated View Systems**
## Practical Assignment 4 - INFOB2DA
*Tobias Buiten & Grace Gunderson*


In [ ]:
# Download Dataset from GitHub Release & Unzip Large Dataset
!curl -L -o DelayedFlights.zip "https://github.com/glgunderson/INFOB2DA-PA4/releases/download/PA4.DATA/DelayedFlights.zip"
!unzip -o DelayedFlights.zip -d data

import pandas as pd

# Load Dataset
df_raw = pd.read_csv('data/airlinedelaycauses_DelayedFlights.csv')

# Copy Dataset for Preprocessing
df = df_raw.copy()

# Preview Dataset
df_raw.head()

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 62.0M  100 62.0M    0     0  54.9M      0  0:00:01  0:00:01 --:--:--  270M
Archive:  DelayedFlights.zip
  inflating: data/airlinedelaycauses_DelayedFlights.csv  


Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  

[5 rows x 30 columns]

## Dataset Overview

### DOT’S Air Travel Consumer Report:
The U.S. Department of Transportation’s (DOT) Bureau of Transportation Statistics (BTS) tracks the **on-time performance** of domestic flights operated by large air carriers.
- DOT provides **monthly summary information** on the number of on-time, delayed, cancelled and diverted flights.
- BTS collects details on the **causes of flight delays** and releases summary statistics and raw data.


## Summary Statistics

In [ ]:
# Understand the dataset
df_raw.info()
df_raw.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

(1936758, 30)

In [ ]:
# Basic Summary Statistics
df_raw.describe().T

count          mean           std     min        25%  \
Unnamed: 0         1936758.0  3.341651e+06  2.066065e+06     0.0  1517452.5   
Year               1936758.0  2.008000e+03  0.000000e+00  2008.0     2008.0   
Month              1936758.0  6.111106e+00  3.482546e+00     1.0        3.0   
DayofMonth         1936758.0  1.575347e+01  8.776272e+00     1.0        8.0   
DayOfWeek          1936758.0  3.984827e+00  1.995966e+00     1.0        2.0   
DepTime            1936758.0  1.518534e+03  4.504853e+02     1.0     1203.0   
CRSDepTime         1936758.0  1.467473e+03  4.247668e+02     0.0     1135.0   
ArrTime            1929648.0  1.610141e+03  5.481781e+02     1.0     1316.0   
CRSArrTime         1936758.0  1.634225e+03  4.646347e+02     0.0     1325.0   
FlightNum          1936758.0  2.184263e+03  1.944702e+03     1.0      610.0   
ActualElapsedTime  1928371.0  1.333059e+02  7.206007e+01    14.0       80.0   
CRSElapsedTime     1936560.0  1.343027e+02  7.134144e+01   -25.0       82.0   
AirTime            1928371.0  1.082771e+02  6.864261e+01     0.0       58.0   
ArrDelay           1928371.0  4.219988e+01  5.678472e+01  -109.0        9.0   
DepDelay           1936758.0  4.318518e+01  5.340250e+01     6.0       12.0   
Distance           1936758.0  7.656862e+02  5.744797e+02    11.0      338.0   
TaxiIn             1929648.0  6.812975e+00  5.273595e+00     0.0        4.0   
TaxiOut            1936303.0  1.823220e+01  1.433853e+01     0.0       10.0   
Cancelled          1936758.0  3.268348e-04  1.807562e-02     0.0        0.0   
Diverted           1936758.0  4.003598e-03  6.314722e-02     0.0        0.0   
CarrierDelay       1247488.0  1.917940e+01  4.354621e+01     0.0        0.0   
WeatherDelay       1247488.0  3.703571e+00  2.149290e+01     0.0        0.0   
NASDelay           1247488.0  1.502164e+01  3.383305e+01     0.0        0.0   
SecurityDelay      1247488.0  9.013714e-02  2.022714e+00     0.0        0.0   
LateAircraftDelay  1247488.0  2.529647e+01  4.205486e+01     0.0        0.0   

                         50%         75%        max  
Unnamed: 0         3242558.0  4972466.75  7009727.0  
Year                  2008.0     2008.00     2008.0  
Month                    6.0        9.00       12.0  
DayofMonth              16.0       23.00       31.0  
DayOfWeek                4.0        6.00        7.0  
DepTime               1545.0     1900.00     2400.0  
CRSDepTime            1510.0     1815.00     2359.0  
ArrTime               1715.0     2030.00     2400.0  
CRSArrTime            1705.0     2014.00     2400.0  
FlightNum             1543.0     3422.00     9742.0  
ActualElapsedTime      116.0      165.00     1114.0  
CRSElapsedTime         116.0      165.00      660.0  
AirTime                 90.0      137.00     1091.0  
ArrDelay                24.0       56.00     2461.0  
DepDelay                24.0       53.00     2467.0  
Distance               606.0      998.00     4962.0  
TaxiIn                   6.0        8.00      240.0  
TaxiOut                 14.0       21.00      422.0  
Cancelled                0.0        0.00        1.0  
Diverted                 0.0        0.00        1.0  
CarrierDelay             2.0       21.00     2436.0  
WeatherDelay             0.0        0.00     1352.0  
NASDelay                 2.0       15.00     1357.0  
SecurityDelay            0.0        0.00      392.0  
LateAircraftDelay        8.0       33.00     1316.0

The initial output upon loading the *full raw dataset* includes:
- **1,936,758 rows (flights) x 30 columns (features)**
- The columns consist of both numeric (`int64`, `float64`) and categorical (`object`) features, including:
  - 14 float variables (e.g., `DepTime`, `ArrTime`, `DepDelay`, `ArrDelay`)
  - 11 integer variables (e.g., `Year`, `Month`, `DayofWeek`, `FlightNum`)
  - 5 object variables (e.g., `UniqueCarrier`, `Origin`, `Dest`)  

According to PA4, the dataset *should* include:
- Flight delay metrics for **1,247,486** different flights.
- **30 different features**, both numerical and categorical.

### Understanding the Record Discrepancy
The difference between the ~1.94 million and ~1.25 million flight records is explained by *dataset scope*.
- The full raw dataset (**1,936,758 rows**) includes **all scheduled flights** in 2008 - whether they were on time, delayed, cancelled, or diverted.
- Only a subset of the flight records (**1,247,488 rows**) contain complete **delay-related data** (`CarrierDelay`, `WeatherDelay`, `NASDelay`, `SecurityDelay`, `LateAircraftDelay`).
  - These represent flights that actually experienced a *delay event*, which is the primary focus of this data analysis.

As a result, all preprocessing and subsequent visualizations are performed on this ~1.25M delayed-flight subset to ensure meaningful data analysis.

In [ ]:
# PREPROCESSING

# Drop redundant/irrelevant columns
# 'Unnamed: 0' = index column auto-generated at export
# 'Year' = constant value (2008)
df = df.drop(columns=['Unnamed: 0', 'Year'])

# Remove cancelled or diverted flights
# These do not have valid arrival/departure data
df = df[(df['Cancelled'] == 0) & (df['Diverted'] == 0)]

# Identify records with complete delay-cause data
# Isolates the ~1.25M delayed-flight subset
delayed = ['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']
df = df.dropna(subset=delayed)

# Drop rows missing essential delay information
# Ensures valid numerical data for both departure and arrival delays
df = df.dropna(subset=['DepDelay', 'ArrDelay'])

# Fill missing delay-cause values with 0
# NaN indicates "no delay" for that cause
delay_cols = ['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']
df[delay_cols] = df[delay_cols].fillna(0)

# Clip negative delay values (representing early arrivals) to 0
# Negative delays imply early arrivals, not true delays
df['DepDelay'] = df['DepDelay'].clip(lower=0)
df['ArrDelay'] = df['ArrDelay'].clip(lower=0)

# Derive new time-based features for later visualization
# Converting scheduled and actual times to hourly bins
df['DepHour'] = (df['DepTime'] // 100).astype(int)
df['ArrHour'] = (df['ArrTime'] // 100).astype(int)

# Reset DataFrame index to ensure clean row alignment
df = df.reset_index(drop=True)

# Verify structure after preprocessing
df.info()
df.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1247488 entries, 0 to 1247487
Data columns (total 30 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   Month              1247488 non-null  int64  
 1   DayofMonth         1247488 non-null  int64  
 2   DayOfWeek          1247488 non-null  int64  
 3   DepTime            1247488 non-null  float64
 4   CRSDepTime         1247488 non-null  int64  
 5   ArrTime            1247488 non-null  float64
 6   CRSArrTime         1247488 non-null  int64  
 7   UniqueCarrier      1247488 non-null  object 
 8   FlightNum          1247488 non-null  int64  
 9   TailNum            1247486 non-null  object 
 10  ActualElapsedTime  1247488 non-null  float64
 11  CRSElapsedTime     1247488 non-null  float64
 12  AirTime            1247488 non-null  float64
 13  ArrDelay           1247488 non-null  float64
 14  DepDelay           1247488 non-null  float64
 15  Origin             1247488 non-n

(1247488, 30)

### Preprocessing Overview
- The first column, **Unnamed: 0**, is an index column automatically generated during export and does not represent a meaningful feature. It will will be removed during preprocessing.

## Analyze Dataset

In [ ]:
# Experiment with basic charts/visualizations

# **INTERACTIVE VISUALIZATION DASHBOARD**

## **Visualization 1**

In [ ]:
# 1

### Conclusion

## **Visualization 2**

In [ ]:
# 2

### Conclusion

## **Visualization 3**

In [ ]:
# 3

### Conclusion

## **Visualization 4**

In [ ]:
# 4

### Conclusion



## **Visualization 5**

In [ ]:
# 5

### Conclusion